fan_control.cpp

```diff
@@ -0,0 +1,194 @@
+                    if ( mode == 0x01 ) { mode_set = QString("low"); }
+                    if ( mode == 0x02 ) { mode_set = QString("medium"); }
+                    if ( mode == 0x03 ) { mode_set = QString("high"); }
+                    if ( mode == 0x04 ) { mode_set = QString("on"); }
+                    if ( mode == 0x05 ) { mode_set = QString("auto"); }
+                    if ( mode == 0x06 ) { mode_set = QString("smart"); }
+
+                    item = sensor->item(RConfigFanMode);
+                    if (item && !item->toString().isEmpty() && item->toString() != mode_set)
+                    {
+                        item->setValue(mode_set);
+                        enqueueEvent(Event(RSensors, RConfigFanMode, sensor->id(), item));
+                        configUpdated = true;
+                    }
+                }
+                sensor->setZclValue(updateType, ind.srcEndpoint(), FAN_CONTROL_CLUSTER_ID, attrId, attr.numericValue());
+            }
+                break;
+
+            default:
+                break;
+            }
+        }
+
+        if (stateUpdated)
+        {
+            sensor->updateStateTimestamp();
+            enqueueEvent(Event(RSensors, RStateLastUpdated, sensor->id()));
+        }
+
+        if (configUpdated || stateUpdated)
+        {
+            updateEtag(sensor->etag);
+            updateEtag(gwConfigEtag);
+            sensor->setNeedSaveDatabase(true);
+            queSaveDb(DB_SENSORS, DB_SHORT_SAVE_DELAY);
+        }
+    }
+}
+
+/*! Write Attribute on fan control cluster.
+   \param task - the task item
+   \param attrId
+   \param attrType
+   \param attrValue
+   \return true - on success
+           false - on error
+ */
+bool DeRestPluginPrivate::addTaskFanControlReadWriteAttribute(TaskItem &task, uint8_t readOrWriteCmd, uint16_t attrId, uint8_t attrType, uint32_t attrValue, uint16_t mfrCode)
+{
+    if (readOrWriteCmd != deCONZ::ZclReadAttributesId && readOrWriteCmd != deCONZ::ZclWriteAttributesId)
+    {
+        DBG_Printf(DBG_INFO, "Thermostat invalid parameter readOrWriteCmd %d\n", readOrWriteCmd);
+        return false;
+    }
+
+    task.taskType = TaskThermostat;
+
+    task.req.setClusterId(FAN_CONTROL_CLUSTER_ID);
+    task.req.setProfileId(HA_PROFILE_ID);
+
+    task.zclFrame.payload().clear();
+    task.zclFrame.setSequenceNumber(zclSeq++);
+    task.zclFrame.setCommandId(readOrWriteCmd);
+    task.zclFrame.setFrameControl(deCONZ::ZclFCProfileCommand |
+            deCONZ::ZclFCDirectionClientToServer |
+            deCONZ::ZclFCDisableDefaultResponse);
+
+    if (mfrCode != 0x0000)
+    {
+        task.zclFrame.setFrameControl(task.zclFrame.frameControl() | deCONZ::ZclFCManufacturerSpecific);
+        task.zclFrame.setManufacturerCode(mfrCode);
+    }
+
+    // payload
+    QDataStream stream(&task.zclFrame.payload(), QIODevice::WriteOnly);
+    stream.setByteOrder(QDataStream::LittleEndian);
+
+    stream << (quint16) attrId;
```

Comment:
Just a note for the future: We don't have a coding style yet, but always use c++ named casts instead old C style casts.

`static_cast<quint16>(attrId)`

This might seam a bit superfluous at first. But soon the next generation code will get proper types for certain values like `ZclAttributeId_t`, `EndpointId_t`, etc to prevent common mistakes.

The named casts help the compiler to figure out invalid type conversions at compile time to prevent accidental errors.

For example:

```
ZclAttributeId_t attrId(0x0077);

stream << (quint8) attrId;              // programmer mistake won't be detected by compiler
stream << static_cast<quint8>(attrId);  // compile error no valid conversion from ZclAttributeId_t to quint8
stream << static_cast<quint16>(attrId); // ok
stream << attrId;                       // ok, because operator<< for QDataStream
```

So once we have the new types we only need to change the function signature:

    addTaskFanControlReadWriteAttribute(TaskItem &task, uint8_t readOrWriteCmd, uint16_t attrId, uint8_t attrType, uint32_t attrValue, uint16_t mfrCode)

to

    addTaskFanControlReadWriteAttribute(TaskItem &task, uint8_t readOrWriteCmd, ZclAttributeId_t attrId, ZclDatatypeId_t attrType, uint32_t attrValue, ZclManufacturerCode_t mfrCode)

If we already use C++ named casts we get a hole lot of error checking without needing to rewrite all those C style casts, searching for them is also quite hard/impossible.